lib/changesreader.js

```diff
@@ -0,0 +1,280 @@
+const stream = require('stream')
+const EVENT_BATCH = 'batch'
+const EVENT_CHANGE = 'change'
+const EVENT_SEQ = 'seq'
+const EVENT_ERROR = 'error'
+
+// streaming line breaker
+const liner = () => {
+  const liner = new stream.Transform({ objectMode: true })
+
+  liner._transform = function (chunk, encoding, done) {
+    let data = chunk.toString('utf8')
+    if (this._lastLineData) {
+      data = this._lastLineData + data
+      this._lastLineData = null
+    }
+
+    const lines = data.split(/\s*\n/)
+    this._lastLineData = lines.splice(lines.length - 1, 1)[0]
+    lines.forEach(this.push.bind(this))
+    done()
+  }
+
+  liner._flush = function (done) {
+    this.push(this._lastLineData)
+    this._lastLineData = null
+    done()
+  }
+
+  return liner
+}
+
+// streaming change processor
+const changeProcessor = (ee, batchSize) => {
+  const changeProcessor = new stream.Transform({ objectMode: true })
+  const buffer = []
+  changeProcessor.lastSeq = '0'
+
+  changeProcessor._transform = function (chunk, encoding, done) {
+    // remove last char from string
+    if (chunk[chunk.length - 1] === ',') {
+      chunk = chunk.slice(0, -1)
+    }
+
+    try {
+      const j = JSON.parse(chunk)
+      buffer.push(j)
+      if (buffer.length >= batchSize) {
+        ee.emit(EVENT_BATCH, buffer.splice(0, batchSize))
+      }
+      done()
+    } catch (e) {
+      // look for last_seq
+      const match = chunk.match(/"last_seq":"([^"]+)"/)
+      if (match) {
+        changeProcessor.lastSeq = match[1]
+      }
+      done()
+    }
+  }
+
+  changeProcessor._flush = function (done) {
+    if (buffer.length > 0) {
+      ee.emit(EVENT_BATCH, buffer.splice(0, buffer.length))
+    }
+    done()
+  }
+
+  return changeProcessor
+}
+
+/**
+ * Monitors the changes feed (after calling .start()/.get()) and emits events
+ *  - EVENT_CHANGE - per change
+ *  - EVENT_BATCH - per batch of changes
+ *  - EVENT_SEQ - per change of sequence number
+ *  - EVENT_ERROR - per 4xx error (except 429)
+ *
+ * @param {String} db - Name of the database.
+ * @param {Function} request - Nano.relax
+ */
+class ChangesReader {
+  // constructor
+  constructor (db, request) {
+    this.db = db
+    this.setDefaults()
+    this.request = request
+  }
+
+  // set defaults
+  setDefaults () {
+    this.ee = new EventEmitter()
+    this.batchSize = 100
+    this.fastChanges = false
+    this.since = 'now'
+    this.includeDocs = false
+    this.timeout = 60000
+    this.started = false
+    this.wait = false
+    this.stopOnEmptyChanges = false // whether to stop polling if we get an empty set of changes back
+    this.continue = true // whether to poll again
+    this.qs = {} // extra querystring parameters
+    this.selector = null
+  }
+
+  // prevent another poll happening
+  stop () {
+    this.continue = false
+  }
+
+  // sleep, promise style
+  async sleep (t) {
+    return new Promise((resolve, reject) => {
+      setTimeout(() => { resolve() }, t)
```

Comment:
Please consider using `resolve` to avoid an unnecessary instantiation:

```js
return new Promise((resolve) => {
  setTimeout(resolve, t)
})
``

Reply:
👍